Expands modular computer terminals

NataKilar opened the pull request:
Greatly expands modular computer terminals in preparation for hacking, and for fun shenanigans in the interim

* Adds terminal commands for file manipulation: Listing, moving, copying, renaming, and deletion.
* Adds network device commands which expose public variables and methods on machinery to terminal commands. These can be added through the network device UI, or through terminal commands themselves. The player facing terminology is "alias" for the name of the command, and "reference" for the variable or method that is modified/called by the network command.
* Adds network receivers, to expose non-network machinery to the network. Network locks have been made a subtype of network receivers.
* Tweaked a few old terminal commands to make them more expandable.

Network device commands are analogous to radio signals, and the code in general is similar. The sanitization of arguments for commands in particular needs another set of eyes on it.

## Changelog
:cl:
feature: Added terminal commands for file manipulation.
feature: Added network commands, which allow for manipulation of certain public vars and methods on network devices via terminal commands
feature: Added network receiver stock parts for exposing non-network machinery to network commands.
/:cl:

afterthought2 commented on code/game/machinery/_machines_base/stock_parts/network_receiver.dm:
```diff
@@ -0,0 +1,41 @@
+
+/obj/item/stock_parts/network_receiver/attackby(obj/item/I, mob/user)
+	. = ..()
+	if(isMultitool(I))
```
Should really be a multitool extension. You may be able to generalize the radio component one, as it handles target acquisition to some extent.

NataKilar replied:
Cut out the multicool configuration entirely per our discussion on discord.

afterthought2 commented on code/game/machinery/_machines_base/stock_parts/network_receiver.dm:
```diff
@@ -0,0 +1,41 @@
+		if(!istype(input) || !CanPhysicallyInteractWith(usr, src))
+			return
+		var/datum/extension/network_device/stock_part/D = get_extension(src, /datum/extension/network_device/)
+		D.machine = weakref(input)
```
Why set it here and not just automatically on install? It's not like there's further configuration that can be done here as far as I can tell.

afterthought2 commented on code/modules/atmospherics/components/binary_devices/pump.dm:
```diff
@@ -154,6 +154,10 @@ Thus, the two variables affect pump operation are set in New():
 
+/obj/machinery/atmospherics/binary/pump/update_use_power(new_use_power)
+	. = ..()
+	queue_icon_update()
```
This should be called automatically if it matters.

MistakeNot4892 replied:
Unsure from author reply if this has been discussed/addressed.

NataKilar replied:
Without this, it doesn't automatically update the icon on power change. You can see the pump manually calls update_icon() in its Topic proc below, too. Not sure if there's an automatic system in place for this already that's just not getting called.

afterthought2 replied:
Oh yeah, this was not the proc I thought it was. This one you have to do manually.

afterthought2 commented on code/modules/modular_computers/networking/device_types/_network_device.dm:
```diff
@@ -20,6 +24,10 @@
 	
+	if(has_commands)
+		LAZYINITLIST(command_and_call)
+		LAZYINITLIST(command_and_write)
```
Why?

afterthought2 commented on code/modules/modular_computers/networking/device_types/_network_device.dm:
```diff
@@ -196,6 +226,181 @@
+					if("TRUE")
+						return TRUE
+					if("FALSE")
+						return FALSE
```
So the write setters should really be doing this already, at least if you use the safe call.

NataKilar replied:
Left in for now, I'll go through with a fine-tooth comb and make sure all the public vars are sanitizing as necessary and delete this.

afterthought2 reviewed (commented):
I think you should auto-add all available vars and methods with e.g. "name" as alias on install.

This is pretty different from radio, in that for radio the tags -> vars/methods mapping are triggers for radio signals, whereas here the alias is just a convenience for the user. If I were to implement it I would have just had fixed aliases as names, and you can listcom with args (or similar) to obtain a man entry with e.g the name, the desc, the format flag, and the read-only status. Setting alias seems more like a power user feature.

Should also allow you to read vars from the command line; haven't read far enough to see if that's the case but doesn't seem to be.

NataKilar commented:
Adjusted some things re: discussion on discord. Command configuration is solely through the command line now, and listcom can now print out a description of commands fed into it a la man. Unskilled users who attempt to call commands will now call a random method command on a cooldown. This isn't exactly engaging, but it works.

Also added a network scanner for getting network information from nearby network devices.

MistakeNot4892 commented on code/game/machinery/doors/airlock.dm:
```diff
@@ -988,11 +988,6 @@ About the new airlock wires panel:
 /obj/machinery/door/airlock/proc/toggle_lock(var/forced = 0)
 	return locked ? unlock() : lock()
 
-/obj/machinery/door/airlock/allowed(mob/M)
```
What does this removal relate to?

NataKilar replied:
Because commands check access, this bit of code meant if you bolted a door via a command you could not then unbolt it, since the access check would always return false. As far as I can tell there's no other reason it should return false when bolted regardless, so I removed it.

afterthought2 replied:
It's probably better to have network device datums use `check_access`, which is the more modern and saner version of `allowed`. They do the same thing except on some specific objs (like doors, here) where `allowed` carries legacy connotations I'm not familiar with.

MistakeNot4892 commented on code/modules/modular_computers/terminal/terminal.dm:
```diff
@@ -33,6 +54,24 @@
 /datum/terminal/Process()
+	if(current_move)
+		var/result = current_move.update_progress()
+		var/error
```
You don't need to declare/set this, just call `append_to_history()` in the if/else.

MistakeNot4892 commented on code/modules/modular_computers/terminal/terminal.dm:
```diff
@@ -4,20 +4,41 @@
 	computer = null
+	current_disk = null
+	for(var/D in disks)
+		var/datum/file_storage/FS = disks[D]
```
Couldn't you use QDEL_NULL_LIST() for this?

MistakeNot4892 commented on code/modules/modular_computers/terminal/terminal.dm:
```diff
@@ -4,20 +4,41 @@
 		computer.terminals -= src
 	computer = null
+	current_disk = null
+	for(var/D in disks)
```
Couldn't this just be:
````
	for(var/D in disks)
		qdel(disks[D])
	disks = null
	````

MistakeNot4892 commented:
Could the available commands on terminals be unit tested to make sure they're valid/returning correctly/not hitting a protected variable somehow? Makes me a tiny bit uneasy.

afterthought2 commented on code/modules/modular_computers/networking/device_types/_network_device.dm:
```diff
@@ -196,6 +221,156 @@
+		if(command_args) // Write to a var.
+			command_args = sanitize_command_args(command_args, variable.var_type)
+			if(isnull(command_args)) // Command could not be converted into a form appropriate for the variable.
+				return "Improper argument type; correct type is [variable.var_type]"
```
It's possible a null literal is a valid argument in some cases, though am not sure. Just be careful if you move this off to vars, as e.g. radio stuff may be passing nulls into num or text stuff.

NataKilar replied:
Looking through the writable vars, it doesn't look like any take in null as a possible value. The actual public variable var-type checking I added should still permit it, though, if the var-type is IC_FORMAT_ANYTHING

afterthought2 commented on code/modules/modular_computers/networking/device_types/_network_device.dm:
```diff
@@ -196,6 +221,156 @@
+		return "Access denied"
+	var/rand_alias = pick(command_and_call)
+	var/decl/public_access/public_method/rand_method = command_and_call[rand_alias]
+	rand_method.perform(get_command_target(), list())
```
Kind of weird to pass an empty list, no?

afterthought2 commented on code/modules/modular_computers/networking/device_types/stock_part.dm:
```diff
@@ -0,0 +1,20 @@
+		return FALSE
+	var/atom/H = holder
+	var/obj/M = H.loc
+	return M.allowed(user)
```
Yeah, here I recommend `check_access` (and on the base version of this proc).

comma commented on maps/tradeship/backup/tradeship-2.dmm:
```diff
@@ -0,0 +1,788 @@
+"aa" = (/turf/space,/area/space)
```
probably shouldn't commit backup maps

comma commented on code/modules/power/fission/core.dm:
```diff
@@ -0,0 +1,182 @@
+#define ACTIVE_THRESHOLD             200
```
is all the fission stuff intended to be in this PR? I see some related bits where controller is, but that's a lot of fissionstuff so unsure

NataKilar replied:
Nope, my mistake. Accidentally staged the files when I was rebasing. Should be fixed now.

comma commented on code/modules/modular_computers/terminal/terminal_commands.dm:
```diff
@@ -223,3 +241,479 @@ Subtypes
+	var/list/rename_args = get_arguments(text)
+
+	if(length(rename_args) < 2)
+		return "rename: Improper syntax, use move \[file name\] \[new name\]."
```
>use move 

rename?
